File: solutions/98.validate-binary-search-tree.py
#
# @lc app=leetcode id=98 lang=python3
#
# [98] Validate Binary Search Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def isValidSubtree(node, low_limit, high_limit):
            if not node:
                return True
            if not low_limit < node.val < high_limit:
                return False

            return isValidSubtree(node.left, low_limit, node.val) and isValidSubtree(node.right, node.val, high_limit)

        return isValidSubtree(root, float("-inf"), float("inf"))

# @lc code=end

File: solutions/98.validate-binary-search-tree.java
/*
 * @lc app=leetcode id=98 lang=java
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidSubtree(root, null, null);
    }

    private boolean isValidSubtree(TreeNode node, Integer minBoundary, Integer maxBoundary) {
        if (node == null) return true;

        if ((maxBoundary != null && node.val >= maxBoundary)
         || (minBoundary != null && node.val <= minBoundary)) {
            return false;
        }

        return isValidSubtree(node.left, minBoundary, node.val) && isValidSubtree(node.right, node.val, maxBoundary);
    }
}
// @lc code=end

